### PR TITLE
Launch fuel and spark cut issue

### PR DESCRIPTION
**Pull request by opelpanfan**
Fix to Launch Spark and Fuel Cut.

Two statements contradicts each other 
!limitedSpark && !limitedFuel
and 
limitedSpark and limitedFuel will never be true to match &=

### CODE REVIEW

**Comment by rusefillc** [comment]
@opelpanfan please clarify 'draft' state

**Review by mck1117 — APPROVED** [review]

**Comment by mck1117** [comment]
I would love a test that fails before this PR, and passes with it.

**Comment by rusefillc** [comment]
Oh I would totally love to see the test as well

@opelpanfan any chance to have a separate PR with a test which proves wrong behavoir?

**Comment by mck1117** [comment]
One option is to move this logic in to `limp_manager.cpp` instead, where we have already validated that it works. We would just have to then write a very simple test to check that limp_manager.cpp requests cut in case of launch.

**Comment by opelpanfan** [comment]
Sorry I'll take a look at how to write tests. Tested on hardware before pushing PR.

**Comment by opelpanfan** [comment]
Done code move to limp_manager.cpp

Sorry dont know how to write tests, but tested on HW.

LimpState LimpManager::allowInjection() const {
	if (!m_allowInjection) {
		return {false, m_allowInjection.clearReason};
	}
	if (!m_transientAllowInjection) {
		return {false, m_transientAllowInjection.clearReason};
	}
	if (engine->launchController.isLaunchFuelRpmRetardCondition()) {
		return {false, ClearReason::LaunchFuel};
	}
	return {true, ClearReason::None};
}

LimpState LimpManager::allowIgnition() const {
	if (!m_allowIgnition) {
		return {false, m_allowIgnition.clearReason};
	}
	if (!m_transientAllowIgnition) {
		return {false, m_transientAllowIgnition.clearReason};
	}	
	if (engine->launchController.isLaunchSparkRpmRetardCondition()) {
		return {false, ClearReason::LaunchSpark};
	}
	return {true, ClearReason::None};
}


Added two new ClearReason's 
enum class ClearReason : uint8_t {
	None, // 0
	Fatal,
	Settings,
	HardLimit, // 3
	FaultRevLimit,
	BoostCut, // 5
	OilPressure,
	StopRequested, // 7
	EtbProblem, // 8
	LaunchSpark, // 9
	LaunchFuel // 10
};

**Comment by mck1117** [comment]
> LaunchSpark, // 9
> LaunchFuel // 10

Don't need two, we can use the same code for both. We separately log fuel/ign cut reason.

**Comment by mck1117** [comment]
> LimpState LimpManager::allowIgnition() const {

please do this in `updateState`

**Comment by opelpanfan** [comment]
ok doing now and update PR

File: firmware/controllers/limp_manager.cpp [hunk 1]
```diff
@@ -75,6 +75,17 @@ todo AndreiKA this change breaks 22 unit tests?
 		allowSpark.clear();
 */
 	}
+	
+	// Fuel cut if launch control engaged
+	if (engine->launchController.isLaunchFuelRpmRetardCondition()) {
```
Comment (mck1117) [inline firmware/controllers/limp_manager.cpp:80]: do we need to check `isLaunchCondition` too?

Comment (mck1117) [reply firmware/controllers/limp_manager.cpp:80]: nope, looks like these functions check that internally. Sweet!

Comment (opelpanfan) [reply firmware/controllers/limp_manager.cpp:80]: it is already checked as part of bool LaunchControlBase::isLaunchRpmRetardCondition()

File: firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp [hunk 2]
```diff
@@ -25,10 +25,10 @@ const char * getBoardSpecificPinName(brain_pin_e brainPin) {
 		case GPIOD_1: return "12 - Ignition 4";
 		case GPIOB_1: return "40 - AN volt 9, Aux Reuse";
 		case GPIOB_0: return "36 - AN volt 8, Aux Reuse";
-		case GPIOB_9: return "AUX J2 PB9";
 		case GPIOD_7: return "14 - GP Out 5";
-		case GPIOB_8: return "AUX J2 PB8";
+		case GPIOB_9: return "AUX J2 PB9";
```
Comment (rusefillc) [inline firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: GRRRRRR

I am very concerned when people do not care what they include in PRs :(

File: firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp [hunk 3]
```diff
@@ -25,10 +25,10 @@ const char * getBoardSpecificPinName(brain_pin_e brainPin) {
 		case GPIOD_1: return "12 - Ignition 4";
 		case GPIOB_1: return "40 - AN volt 9, Aux Reuse";
 		case GPIOB_0: return "36 - AN volt 8, Aux Reuse";
-		case GPIOB_9: return "AUX J2 PB9";
```
Comment (rusefillc) [inline firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:28]: this change does not belong in this PR

**Review by rusefillc — CHANGES_REQUESTED** [review]

File: firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp (hunk 2 above)
Comment (opelpanfan) [reply firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: I don't think it was me.

Comment (opelpanfan) [reply firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: I haven't pushed this on.

[Auto-generated configs and docs](https://github.com/rusefi/rusefi/pull/3988/commits/0ce9438397c00070dcee53641612e8d447f94404)

Comment (mck1117) [reply firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: you did push it, because you mis-merged it from master

Comment (rusefillc) [reply firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: You are totally right I am just being defensive and aggressive by default please forgive me, some personal drama today

https://github.com/rusefi/rusefi/issues/3989

Comment (opelpanfan) [reply firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: how do I fix this?

Comment (opelpanfan) [reply firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: change back to         case GPIOB_8: return "AUX J2 PB8";

Comment (rusefillc) [reply firmware/config/boards/microrusefi/connectors/generated_ts_name_by_pin.cpp:29]: considering the value of this fix here and identifying https://github.com/rusefi/rusefi/issues/3989 we can just ignore the imperfection it and merge as is

File: firmware/controllers/limp_manager.cpp [hunk 4]
```diff
@@ -140,6 +151,6 @@ LimpState LimpManager::allowIgnition() const {
 	}
 	if (!m_transientAllowIgnition) {
 		return {false, m_transientAllowIgnition.clearReason};
-	}
+	}
```
Comment (rusefillc) [inline firmware/controllers/limp_manager.cpp:154]: @opelpanfan mostly because I am a grumpy old man. Why did we need to change this line?

Please consider reviewing every line you push/PR.

File: firmware/controllers/limp_manager.h [hunk 5]
```diff
@@ -13,7 +13,8 @@ enum class ClearReason : uint8_t {
 	BoostCut, // 5
 	OilPressure,
 	StopRequested, // 7
-	EtbProblem,
+	EtbProblem, // 8
+	LaunchCut // 9
```
Comment (rusefillc) [inline firmware/controllers/limp_manager.h:17]: please add a comma so that next person changing here is not changing your line
